### PR TITLE
Contract and harness for copy_to, copy_to_nonoverlapping, copy_from, and copy_from_nonoverlapping

### LINKED ISSUE #53: Challenge 6: Safety of `NonNull`
[Challenge 6: Safety of `NonNull`](https://model-checking.github.io/verify-rust-std/challenges/0006-nonnull.html#challenge-6-safety-of-nonnull)

### PR DESCRIPTION
**Pull request by Dhvani-Kapadia**
Description

This PR includes contracts and proof harnesses for the four APIs copy_to, copy_to_nonoverlapping, copy_from, and copy_from_nonoverlapping which are part of the NonNull library in Rust.

Changes Overview:

Covered APIs:
NonNull::copy_to
NonNull::copy_to_nonoverlapping
NonNull::copy_from
NonNull::opy_from_nonoverlapping

Proof harness:
non_null_check_copy_to
non_null_check_copy_to_nonoverlapping
non_null_check_copy_from
non_null_check_copy_from_nonoverlapping,

Revalidation

To revalidate the verification results, run path_to/kani/scripts/kani verify-std -Z unstable-options "path/to/library" -Z function-contracts -Z mem-predicates --harness ptr::non_null::verify. This will run all four harnesses in the module. All default checks should pass:

SUMMARY:
 ** 0 of 141 failed

VERIFICATION:- SUCCESSFUL
Verification Time: 0.62114185s

Complete - 6 successfully verified harnesses, 0 failures, 6 total.

Towards issue #53 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Review by celinval — COMMENTED** [review]
Thanks!

File: library/core/src/ptr/non_null.rs [hunk 1]
```diff
@@ -1803,4 +1832,106 @@ mod verify {
...
+
+        unsafe {
+            kani::assume(kani::mem::can_dereference(src_u8));
+            kani::assume(kani::mem::can_dereference(
+                src_u8.add(core::mem::size_of::<i32>() - 1),
+            ));
+            kani::assume(kani::mem::can_dereference(dest_u8));
+            kani::assume(kani::mem::can_dereference(
+                dest_u8.add(core::mem::size_of::<i32>() - 1),
+            ));
```
[the remark this answers is not in the document]
Comment (Dhvani-Kapadia) [reply library/core/src/ptr/non_null.rs:1975]: I added these assumptions to handle the safety requirements :'_src must be [valid](https://doc.rust-lang.org/std/ptr/index.html#safety) for reads of count * size_of::<T>() bytes, and must remain valid even when dst is written for count * size_of::<T>() bytes._' for the same this issue was also created [#3672](https://github.com/model-checking/kani/issues/3672l) .

File: library/core/src/ptr/non_null.rs [hunk 2]
```diff
@@ -1803,4 +1832,106 @@ mod verify {
...
+            src.copy_to_nonoverlapping(dest, count);
+        }
+    }
+    #[kani::proof_for_contract(NonNull::<T>::copy_from_nonoverlapping)]
+    pub fn non_null_check_copy_from_nonoverlapping() {
+        let mut src_value: i32 = kani::any();
+        let mut dest_value: i32 = 0;
+
+        let src = NonNull::new(&mut src_value as *mut i32).unwrap();
+        let dest = NonNull::new(&mut dest_value as *mut i32).unwrap();
```
[the remark this answers is not in the document]
Comment (Dhvani-Kapadia) [reply library/core/src/ptr/non_null.rs:1960]: ok thanks I will use it

File: library/core/src/ptr/non_null.rs [hunk 3]
```diff
@@ -934,6 +934,12 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_unstable(feature = "const_intrinsic_copy", issue = "80697")]
+    #[kani::requires(count * core::mem::size_of::<T>() <= isize::MAX as usize)]
```
[the remark this answers is not in the document]
Comment (Dhvani-Kapadia) [reply library/core/src/ptr/non_null.rs:969]: Thanks, I used checked_mul to handle overflow

File: library/core/src/ptr/non_null.rs [hunk 4]
```diff
@@ -934,6 +934,12 @@ impl<T: ?Sized> NonNull<T> {
...
+    #[requires(ub_checks::can_dereference(self.as_ptr()))] //The pointer is valid 
+    #[requires(ub_checks::can_dereference(dest.as_ptr()))]
```
[the remark this answers is not in the document]
Comment (Dhvani-Kapadia) [reply library/core/src/ptr/non_null.rs:971]: Thanks, I have added the can_write clause to the contract

File: library/core/src/ptr/non_null.rs [hunk 5]
```diff
@@ -954,6 +960,14 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_unstable(feature = "const_intrinsic_copy", issue = "80697")]
+    #[kani::requires(count * core::mem::size_of::<T>() <= isize::MAX as usize)]
+    #[requires(ub_checks::can_dereference(self.as_ptr()))]
+    #[requires(ub_checks::can_dereference(dest.as_ptr()))]
+    #[requires((self.as_ptr() as usize).abs_diff(dest.as_ptr() as usize) >= count * core::mem::size_of::<T>())]
+    #[ensures(|result: &()| ub_checks::can_dereference(self.as_ptr()))]
+    #[ensures(|result: &()| ub_checks::can_dereference(dest.as_ptr()))]
```
[the remark this answers is not in the document]
Comment (Dhvani-Kapadia) [reply library/core/src/ptr/non_null.rs:1004]: Sure, i will use it.

File: library/core/src/ptr/non_null.rs [hunk 6]
```diff
@@ -934,6 +934,12 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_unstable(feature = "const_intrinsic_copy", issue = "80697")]
+    #[kani::requires(count * core::mem::size_of::<T>() <= isize::MAX as usize)]
+    #[requires(ub_checks::can_dereference(self.as_ptr()))] //The pointer is valid
```
[the remark this answers is not in the document]
Comment (Dhvani-Kapadia) [reply library/core/src/ptr/non_null.rs:970]: I have used checked_mul along with this: `
#[requires(count.checked_mul(core::mem::size_of::<MaybeUninit<T>>()).map_or(false, |size| size <= isize::MAX as usize))]`

File: library/core/src/ptr/non_null.rs [hunk 7]
```diff
@@ -985,6 +992,14 @@ impl<T: ?Sized> NonNull<T> {
...
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "const_intrinsic_copy", since = "1.83.0")]
+    #[requires(count.checked_mul(core::mem::size_of::<T>()).map_or(false, |size| size <= isize::MAX as usize)
+        && ub_checks::can_dereference(self.as_ptr() as *const MaybeUninit<T>)
+        && ub_checks::can_write(dest.as_ptr())
+        && ub_checks::can_write(dest.as_ptr().add(count))
+        && (self.as_ptr() as usize).abs_diff(dest.as_ptr() as usize) >= count * core::mem::size_of::<T>())]
+        #[ensures(|result: &()| ub_checks::can_dereference(self.as_ptr() as *const u8)
+        && ub_checks::can_dereference(dest.as_ptr() as *const u8))]
+    #[kani::modifies(dest.as_ptr())]
```
Comment (qinheping) [inline library/core/src/ptr/non_null.rs:1002]:
```suggestion
    #[ensures(|result: &()| ub_checks::can_dereference(self.as_ptr() as *const u8)
        && ub_checks::can_dereference(dest.as_ptr() as *const u8))]
    #[kani::modifies(dest.as_ptr())]
```

File: library/core/src/ptr/non_null.rs [hunk 8]
```diff
@@ -1826,7 +1858,68 @@ mod verify {
...
         let xptr = &mut x;
-        let maybe_null_ptr =  if kani::any() { xptr as *mut i32 } else { null_mut() };
+        let maybe_null_ptr = if kani::any() { xptr as *mut i32 } else { null_mut() };
         let _ = NonNull::new(maybe_null_ptr);
     }
+
+    #[kani::proof_for_contract(NonNull::<T>::copy_to)]
+    pub fn non_null_check_copy_to() {
+        // PointerGenerator instance
+        let mut generator = PointerGenerator::<16>::new();
```
Comment (qinheping) [inline library/core/src/ptr/non_null.rs:1868]: How the capacity of the generator affect the running time of the proof? Can we use the maximum capacity, e.g., `isize::MAX / size_of::<i32>()` here?

**Comment by Dhvani-Kapadia** [comment]
This is my contract and harness for fn copy_to,I run into these two failed checks.could you please help me understand where I am going wrong?


```
#[requires(
        count.checked_mul(core::mem::size_of::<T>()).map_or(false, |size| size <= isize::MAX as usize) &&        
ub_checks::can_dereference(self.as_ptr()) &&
       (count == 0 || ub_checks::can_dereference(self.as_ptr().wrapping_add(count - 1))) &&  
       ub_checks::can_write(dest.as_ptr()) &&

        (count == 0 || ub_checks::can_write(dest.as_ptr().wrapping_add(count - 1))) &&

        kani::mem::same_allocation(self.as_ptr(), self.as_ptr().wrapping_add(count - 1)) &&

        kani::mem::same_allocation(dest.as_ptr(), dest.as_ptr().wrapping_add(count - 1))
    )]
    #[ensures(|_: &()| 
        ub_checks::can_dereference(self.as_ptr()) &&
        ub_checks::can_dereference(dest.as_ptr()))]
    #[kani::modifies(self.as_ptr(), dest.as_ptr())]
    pub const unsafe fn copy_to(self, dest: NonNull<T>, count: usize)
    where
        T: Sized,
    {
        // SAFETY: the caller must uphold the safety contract for `copy`.
        unsafe { ptr::copy(self.pointer, dest.as_ptr(), count) }
    }
 
```

```
#[kani::proof_for_contract(NonNull::<T>::copy_to)]
    pub fn non_null_check_copy_to() {
        // Generate arbitrary raw pointers for src and dest
        const SIZE: usize = mem::size_of::<i32>() * 100;
        let mut generator = PointerGenerator::<SIZE>::new();  
        let src_ptr: *mut i32 =  generator.any_in_bounds().ptr as *mut i32;
        let dest_ptr: *mut i32 =  generator.any_in_bounds().ptr as *mut i32;
        let src_nonnull = NonNull::new(src_ptr).unwrap();
        let dest_nonnull = NonNull::new(dest_ptr).unwrap();
        let max_count = SIZE / mem::size_of::<i32>();
        let count: usize = kani::any_where(|&x| x > 0 && x <= max_count);
        unsafe { src_nonnull.copy_to(dest_nonnull, count); }
            }
        }

```
```
check 39: memmove.assigns.2
	 - Status: FAILURE
	 - Description: "Check that array_replace((const void *)(char *)dest, ...) is allowed by the assigns clause"
	 - Location: <builtin-library-memmove>:34 in function memmove //ADD modifies clause

     Check 28: kani::mem::is_inbounds::<(), i32>.assertion.1
	 - Status: FAILURE
	 - Description: "Kani does not support reasoning about pointer to unallocated memory"
	 - Location: library/core/src/lib.rs:421:1 in function kani::mem::is_inbounds::<(), i32> ,
```

**Comment by zhassan-aws** [comment]
@Dhvani-Kapadia the modifies clause:
```rust
#[kani::modifies(dest.as_ptr())]
```
states that only the element that `dest` points to is modified. However, the function modifies the entire region between `dest` and `dest + size_of::<T>*count`. To fix this, the modifies clause needs to be in terms of the entire slice, e.g.
```rust
#[kani::modifies(NonNull:slice_from_raw_parts(dest, count).as_ptr()]
```

File: library/core/src/ptr/non_null.rs (hunk 8 above)
Comment (QinyuanWu) [reply library/core/src/ptr/non_null.rs:2468]: With size 16 the runtime is about one second:
```
Checking harness ptr::non_null::verify::non_null_check_copy_to...

VERIFICATION RESULT:
 ** 0 of 328 failed (4 unreachable)

VERIFICATION:- SUCCESSFUL
Verification Time: 0.9928743s
```
With `usize::MAX / size_of::<i32>()` the value is too big:
```
error: internal compiler error: kani-compiler/src/codegen_cprover_gotoc/context/goto_ctx.rs:336:9: failed to get layout for `kani::arbitrary_ptr::PointerGenerator<4611686018427387903>`: values of the type `[u8; 4611686018427387903]` are too big for the target architecture
```
And with `isize::MAX as usize)/ size_of::<i32>()` the runtime is similar but it resulted in four failed checks:
```
Checking harness ptr::non_null::verify::non_null_check_copy_to...

VERIFICATION RESULT:
 ** 4 of 328 failed (4 unreachable)
Failed Checks: CAR size is less than __CPROVER_max_malloc_size
 File: "<builtin-library-__CPROVER_contracts_library>", line 792, in __CPROVER_contracts_write_set_check_assignment
Failed Checks: no offset bits overflow on CAR upper bound computation
 File: "<builtin-library-__CPROVER_contracts_library>", line 798, in __CPROVER_contracts_write_set_check_assignment
Failed Checks: CAR size is less than __CPROVER_max_malloc_size
 File: "<builtin-library-__CPROVER_contracts_library>", line 164, in __CPROVER_contracts_car_set_insert
Failed Checks: no offset bits overflow on CAR upper bound computation
 File: "<builtin-library-__CPROVER_contracts_library>", line 168, in __CPROVER_contracts_car_set_insert

VERIFICATION:- FAILED
Verification Time: 1.2587996s
```

**Comment by QinyuanWu** [comment]
@zhassan-aws Thank you for the suggestion! I've applied the changes and two failed checks for `copy_to` are resolved.

**Comment by Dhvani-Kapadia** [comment]
Thanks @zhassan-aws, @celinval and @qinheping , 
with Olivias help the two failed checks passed.Could you please review it?

File: library/core/src/ptr/non_null.rs [hunk 9]
```diff
@@ -1021,6 +1021,12 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "const_intrinsic_copy", since = "1.83.0")]
+    #[requires(count.checked_mul(core::mem::size_of::<T>()).map_or(false, |size| size <= isize::MAX as usize)
```
Comment (carolynzech) [inline library/core/src/ptr/non_null.rs:1024]:
```suggestion
    #[requires(count.checked_mul(core::mem::size_of::<T>()).map_or_else(|| false, |size| size <= isize::MAX as usize)
```
`map_or_else` [is recommended](https://doc.rust-lang.org/std/option/enum.Option.html#method.map_or) when passing the result of a function call.

File: library/core/src/ptr/non_null.rs [hunk 10]
```diff
@@ -1041,6 +1047,13 @@ impl<T: ?Sized> NonNull<T> {
...
```
Comment (carolynzech) [inline library/core/src/ptr/non_null.rs:1050]:
```suggestion
    #[requires(count.checked_mul(core::mem::size_of::<T>()).map_or_else(|| false, |size| size <= isize::MAX as usize)
```

File: library/core/src/ptr/non_null.rs [hunk 11]
```diff
@@ -1041,6 +1047,13 @@ impl<T: ?Sized> NonNull<T> {
...
+        && ub_checks::can_dereference(NonNull::slice_from_raw_parts(self, count).as_ptr())
+        && ub_checks::can_write(NonNull::slice_from_raw_parts(dest, count).as_ptr())
+        && (self.as_ptr() as usize).abs_diff(dest.as_ptr() as usize) >= count * core::mem::size_of::<T>())]
```
Comment (carolynzech) [inline library/core/src/ptr/non_null.rs:1053]:
```suggestion
        && ub_checks::maybe_is_nonoverlapping(self.as_ptr() as *const (), dest.as_ptr() as *const (), count, core::mem::size_of::<T>()))]
```
Better not to duplicate functionality in `ub_checks` if we can avoid it.

File: library/core/src/ptr/non_null.rs [hunk 12]
```diff
@@ -1061,6 +1074,12 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "const_intrinsic_copy", since = "1.83.0")]
+    #[requires(count.checked_mul(core::mem::size_of::<T>()).map_or(false, |size| size <= isize::MAX as usize)
```
Comment (carolynzech) [inline library/core/src/ptr/non_null.rs:1077]:
```suggestion
    #[requires(count.checked_mul(core::mem::size_of::<T>()).map_or_else(|| false, |size| size <= isize::MAX as usize)
```

File: library/core/src/ptr/non_null.rs [hunk 13]
```diff
@@ -1081,6 +1100,13 @@ impl<T: ?Sized> NonNull<T> {
...
```
Comment (carolynzech) [inline library/core/src/ptr/non_null.rs:1103]:
```suggestion
    #[requires(count.checked_mul(core::mem::size_of::<T>()).map_or_else(|| false, |size| size <= isize::MAX as usize)
```

File: library/core/src/ptr/non_null.rs [hunk 14]
```diff
@@ -1081,6 +1100,13 @@ impl<T: ?Sized> NonNull<T> {
...
+        && ub_checks::can_dereference(NonNull::slice_from_raw_parts(src, count).as_ptr())
+        && ub_checks::can_write(NonNull::slice_from_raw_parts(self, count).as_ptr())
+        && (self.as_ptr() as usize).abs_diff(self.as_ptr() as usize) >= count * core::mem::size_of::<T>())]
```
Comment (carolynzech) [inline library/core/src/ptr/non_null.rs:1106]: same comment about using `maybe_is_overlapping` instead

File: library/core/src/ptr/non_null.rs [hunk 15]
```diff
@@ -2475,4 +2501,62 @@ mod verify {
...
+         let src_ptr = generator.any_in_bounds::<i32>().ptr;
+         let dest_ptr = generator.any_in_bounds::<i32>().ptr;
+          // Wrap the raw pointers in NonNull
+         let src = NonNull::new(src_ptr).unwrap();
+         let dest = NonNull::new(dest_ptr).unwrap();
+         // Generate an arbitrary count using kani::any
+         let count: usize = kani::any();
+         unsafe { src.copy_from_nonoverlapping(dest, count);}
+    }
+}
```
Comment (carolynzech) [inline library/core/src/ptr/non_null.rs:2562]:
```suggestion
}

```
add new line

**Comment by Dhvani-Kapadia** [comment]
Thanks @celinval , I resolved the conversations.

File: library/core/src/ptr/non_null.rs [hunk 16]
```diff
@@ -1021,6 +1021,12 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "const_intrinsic_copy", since = "1.83.0")]
+    #[requires(count.checked_mul(core::mem::size_of::<T>()).map_or_else(|| false, |size| size <= isize::MAX as usize)
+        && ub_checks::can_dereference(NonNull::slice_from_raw_parts(self, count).as_ptr())
+        && ub_checks::can_write(NonNull::slice_from_raw_parts(dest, count).as_ptr()))]
+    #[ensures(|result: &()| ub_checks::can_dereference(self.as_ptr() as *const u8)
+        && ub_checks::can_dereference(dest.as_ptr() as *const u8))]
+    #[kani::modifies(NonNull::slice_from_raw_parts(dest, count).as_ptr())]
```
Comment (tautschnig) [inline library/core/src/ptr/non_null.rs:1029]: This now fails to compile.

Comment (Dhvani-Kapadia) [reply library/core/src/ptr/non_null.rs:1029]: I just updated the branch with main and does successfully run, could you please check again?
````
Checking harness ptr::non_null::verify::non_null_check_copy_to_nonoverlapping...

VERIFICATION RESULT:
 ** 0 of 377 failed (2 unreachable)

VERIFICATION:- SUCCESSFUL
Verification Time: 6.860433s

Checking harness ptr::non_null::verify::non_null_check_copy_to...

VERIFICATION RESULT:
 ** 0 of 340 failed (4 unreachable)

VERIFICATION:- SUCCESSFUL
Verification Time: 6.714273s

Complete - 2 successfully verified harnesses, 0 failures, 2 total.`
```

Comment (tautschnig) [reply library/core/src/ptr/non_null.rs:1029]: This still fails, see https://github.com/model-checking/verify-rust-std/actions/runs/12267053169/job/34226426202?pr=149.

**Review by carolynzech — APPROVED** [review]

**Pull request merged**